### DISCUSSION TITLE
add NilThen api

### DISCUSSION THREAD

**Discussion by shynome** [lainio/err2#16] [Ideas]
add `err2.NilThen` method and support `try.Out(error).Handle(&err)` bind error to err

I don't know much about ziglang `errdefer` and `defer`, but these api can give us a similar experience in golang
```go
func main() {
	var err error
	logger := slog.With()

	logger.Debug("add node")
	defer err2.NilThen(&err, func() {
		logger.Debug("add node successful")
	})
	defer err2.Handle(&err, func() { try.To(err) })

	try.To(AddNode())
}
```

<details>

<summary>I use the blow code now</summary>

```go
package main

import (
	"github.com/lainio/err2/try"
	"github.com/shynome/err0"
	"golang.org/x/exp/slog"
)

func main() {
	var err error
	logger := slog.With()

	logger.Debug("add node")
	defer err0.NilThen(&err, func() {
		logger.Debug("add node successful")
	})
	defer err2.Handle(&err, func() { try.To(err) })

	try.To(AddNode())
}

func AddNode() error { return nil }
```

</details>

**Comment by shynome** [comment]
I found use `defer err2.Handle(&err, func() { try.To(err) })` is more clearly
```golang
func main() {
	var err error
	logger := slog.With()

	logger.Debug("add node")
	defer err2.NilThen(&err, func() {
		logger.Debug("add node successful")
	})
	defer err2.Handle(&err, func() { try.To(err) })

	try.To(AddNode())
}
```

**Reply by lainio** [reply]
Thank you very much for your ideas and contribution to the discussions.

I'm sorry for not getting back to you sooner. 

I noticed that you had closed this. What are you building or want to achieve? Could I help?

I noticed a couple of things:
1. In the original code block, the line `try.Out(AddNode()).Handle(&err)`, the `&err` argument does nothing. If you check the documentation (which I naturally need to update), that's not required. It gets the current `err` values from the `Result` struct's member, filled by the `try.Out` function. In [the documentation](https://pkg.go.dev/github.com/lainio/err2@v0.9.29/try#example-Out1-CopyFile), there is a `CopyFile` function which shows what's the idea behind the `try.Out` concept. What might be confusing is the fact that the `try.Out` calls still need at least one `defer err2.Handle` at the beginning of the current function.
2. You are using the line `defer err2.Handle(&err, func() { try.To(err) })`. That line does nothing because the closure is called only if `err != nil` and then it `try.To(err)` throws the same error again. All this happens because the `defer err2.Handle()` internally has caught the error using the `recover` built-in function.
3. I checked your `err0.NilThen` implementation, and it doesn't work because it doesn't call the `recover` function. So, it has an 'old' `err` value.

In summary, seeing the correct use case and the need for 'err2.NilThen` is tricky.

**Reply by shynome** [reply]
Thanks your reply, I have changed the title, now I only want api `NilThen`
1. I don't use `try.Out` api anymore. I like `try.To` better
2. I need `defer err2.Handle(&err, func() { try.To(err) })` bind error to err but don't handle it.
3. because 2 if any error has be throw the `*err` will be not nil, when err is not nil `err0.NilThen` don't run the function. 
you are right I can add a `recover` to `NilThen` then I can remove `defer err2.Handle(&err, func() { try.To(err) })`, less code always better, thanks for your suggestion

**Reply by shynome** [reply]
ohhhhh thanks for your advice!
now  the `err0.NilThen` is too simple, just a tiny helper function
```go
package err0

func NilThen(fn func()) {
	if r := recover(); r != nil {
		panic(r)
	}
	fn()
}
```
-----
appand: don't know how about performance, but I will continue use this

**Reply by lainio** [reply]
I tried to figure out where you are using the `NilThen` helper but didn't succeed.

Why you don't just write:
```go
	logger.Debug("1: ADD node")

	try.To(AddNode())
	logger.Debug("1: add node successful")
```

Could you show me your algorithm?

**Reply by shynome** [reply]
when function body has many line, defer log is clear 
```go
	logger.Debug("1: ADD node")
	//defer logger.Debug("1: add node successful")

	try.To(AddNode())
	try.To(AddNode())
	try.To(AddNode())
	try.To(AddNode())
	try.To(AddNode())
	try.To(AddNode())
	logger.Debug("1: add node successful")
```

**Reply by lainio** [reply]
Okay, I understand your case now.

I cannot yet promise a new API, but I will study the subject. The first idea that came to my mind is something like this:

```go
	defer err2.Handle(&err, nil, func() {   // 👈 note the nil in the argument list
		logger.Info("add node")
	})
```
Or a new sub-package like `try` and `assert` are:
```go
	defer noerr.Handle(&err, func() {   // 👈 note the 'noerr', or 'ok', or something else..
		logger.Info("add node")
	})
```
I'm not very fond of any of these. I have to think because I have tried to keep API as simple as possible and even removed functions from it later.

**Reply by shynome** [reply]
how about add a CheckHandler Type, always run it, if err==nil, `noerr` param is true
```go
	defer err2.Handle(&err, func(noerr bool) { // always run check
		if noerr {
			logger.Info("added node")
			return
		}
		logger.Info("add node failed", err)
	})
```

**Reply by lainio** [reply]
Yes, that's better 👍

**Reply by lainio** [reply]
Hi, I'm in the middle of releasing v0.9.40 and wanted to let you know that the feature you have requested and we have discussed here is implemented. I'm still calling it experimental because I have to see by myself that it's really useful:
```go
    defer err2.Handle(&err, func(noerr bool) {
        //The handler is called from defer-stack only and only if the current err value is nil!
        assert.That(noerr) // .. so noerr is always true!! Errors are handled as before
        logger.Debug("3: add node successful")
    })
```
As you can see, it's slightly different than in your example. Because `err2` supports as many `err2.Handle` -error handlers, I decided to implement it like this. Also, as you can see the `noerr` is always `true`, **and you cannot or should not (closure) change the current error value**. This was the only way that I was satisfied with the API semantics. (The err2 package allows an unlimited number of error handlers added to the defer stack.)

**The v0.9.40 is a very significant version** because I optimized the performance up to 3x compared to the previous version. The reason is a long story, but you can run the `err2` benchmark to see if you are interested. This brought the need to change all the handler function signatures to `func(error) error`. *They don't need to be function closures to be able to read and write the current error value*. They now just function, input, and output. 

I cannot rely on a compiler with this change because Go doesn't have function overloading, but I have implemented the same thing by 'hand' (type assertions). The type mistakes are detected on runtime. Please follow the `scripts/README.md` migration guide for v0.9.40. The migration scripts can be used as a linter as well.

**Reply by shynome** [reply]
`err2.Handle(&err, func(noerr bool) {})` is conflict with multi err2.handle
```go
package main

import (
	"errors"
	"log"

	"github.com/lainio/err2"
	"github.com/lainio/err2/try"
)

func main() {
	var err error
	defer func() {
		log.Println("????????", err)
	}()
	defer err2.Handle(&err, func(noerr bool) {
		// here has be executed. it should not be conflict with multi err2.handle or this is expect ?
		log.Println("noerr handle2", noerr, err)
	})
	defer err2.Handle(&err, func(err error) error {
		return err
	})
	try.To(errors.New("wwww"))
}

```

**Reply by shynome** [reply]
```go
package main

import (
	"errors"
	"log"

	"github.com/lainio/err2"
	"github.com/lainio/err2/try"
)

func main() {
	var err error
	defer func() {
		log.Println("????????", err)
	}()
	defer err2.Handle(&err, func(noerr bool) {
		// I still need add a check here, the noerr param make me confuse. or now this have a bug here?
		if err != nil {
			return
		}
		log.Println("noerr handle2", noerr, err)
	})
	defer err2.Handle(&err, func(err error) error {
		return err
	})
	try.To(errors.New("wwww"))
}
```

**Reply by lainio** [reply]
Good catch! That must be a bug (or something harder), which makes it a little bit weird because I made a test for that: `TestHandle_NoError` into `err2`.  But in the test, those handlers are in a different order. The `func(noerr bool)` is the second one. Here is [a playground to test it](https://go-vim.appspot.com/p/Zx54inxQeDg).

Back to the drawing board 😄

**Reply by lainio** [reply]
I found it and wrote a couple of new tests, plus the fix. Can I wait a day or two before releasing the fix, v0.9.41?

In the meanwhile, you can use the handler order trick or something else.

I would like to wait 24 hours if someone else finds something else.

Once again, thank you!

**Reply by shynome** [reply]
Ohhh, you have done it so quickly. 👍
Of course you can, that is good.

Thanks again for your good library

**Reply by shynome** [reply]
I have checked issue-18 branch, I found I call `try.To(errors.New("new error"))` in ErrorHandler after CheckHandler called, the top err is not changed, only when I add a more ErrorHandler the err will be changed

```go
package main

import (
	"errors"
	"log"

	"github.com/lainio/err2"
	"github.com/lainio/err2/try"
)

func main() {
	var err error
	defer func() {
		log.Println("????????", err)
	}()
	defer err2.Handle(&err, func(noerr bool) {
		log.Println("noerr handle2", noerr, err)
	})
	defer err2.Handle(&err)

	// commnet the below code, you will see diffrent logout
	defer err2.Handle(&err, func(err error) error {
		log.Println("aaaaaaaaaaaa")
		return err
	})

	defer err2.Handle(&err, func(err error) error {
		try.To(errors.New("pppppppppppp"))
		return nil
	})

	// commnet the below code, you will see diffrent logout again
	defer err2.Handle(&err, func(noerr bool) {
		log.Println("noerr handle3", noerr, err)
	})

	try.To(errors.New("wwww"))
}
```

**Reply by lainio** [reply]
Thank you! Awesome that you have taken the time to test the branch already. Highly appreciated.

I have now added enough tests to find and fix the bug.

I haven't released the v0.9.41 yet because of some things to consider with the auto-migration scripts. The company I'm working with uses the `err2` in code bases with hundreds of thousands of lines of Go code. So, automation is the only option. Trying to hit two birds with the one rock ;-) 

Regarding the bug, you're pointing out with your example that the `try.To` function is not the proper one to use inside the `err2.Handle` handler. 

**Don't do this**:
```go
	...
	defer err2.Handle(&err, func(err error) error {
		defer err2.Handle(&err) //This line cannot WORK inside the previous error handler
		try.To(errors.New("pppppppppppp")) // Panics in middle of panicking!
		return nil
	})
```
*It must be like this*:
```go
	...
	defer err2.Handle(&err, func(err error) error {
		return errors.New("pppppppppppp") // now we are just overwriting the err value
	})
```
